Source/WTF/wtf/FunctionPtr.h

```diff
@@ -0,0 +1,228 @@
+// decorated and undecorated null.
+#define ASSERT_NULL_OR_VALID_CODE_POINTER(ptr) \
+    ASSERT(!ptr || reinterpret_cast<intptr_t>(ptr) & ~1)
+#else
+#define ASSERT_NULL_OR_VALID_CODE_POINTER(ptr) // Anything goes!
+#endif
+
+template<PtrTag tag, typename Out, typename... In>
+class TypedFunctionPtr {
+public:
+    using Ptr = Out(*)(In...);
+
+    constexpr TypedFunctionPtr() = default;
+    constexpr TypedFunctionPtr(std::nullptr_t) : m_ptr(nullptr) { }
+
+    constexpr TypedFunctionPtr(Out(*ptr)(In...))
+        : m_ptr(encode(ptr))
+    { }
+
+// MSVC doesn't seem to treat functions with different calling conventions as
+// different types; these methods already defined for fastcall, below.
+#if CALLING_CONVENTION_IS_STDCALL && !OS(WINDOWS)
+    constexpr TypedFunctionPtr(Out(CDECL *ptr)(In...))
+        : m_ptr(encode(ptr))
+    { }
+#endif
+
+#if COMPILER_SUPPORTS(FASTCALL_CALLING_CONVENTION)
+    constexpr TypedFunctionPtr(Out(FASTCALL *ptr)(In...))
+        : m_ptr(encode(ptr))
+    { }
+#endif
+
+    Out operator()(In... in) const
+    {
+        ASSERT(m_ptr);
+        return (*get())(std::forward<In>(in)...);
+    }
+
+    constexpr Ptr get() const { return decode(m_ptr); }
+
+    template<PtrTag otherTag>
+    TypedFunctionPtr<otherTag, Out, In...> retagged() const
+    {
+        if (!m_ptr)
+            return TypedFunctionPtr<otherTag, Out, In...>();
+        return TypedFunctionPtr<otherTag, Out, In...>(decode(m_ptr));
+    }
+
+    constexpr void* executableAddress() const { return reinterpret_cast<void*>(m_ptr); }
+
+    template<PtrTag newTag>
+    void* retaggedExecutableAddress() const { return retagCodePtr<void*, tag, newTag>(m_ptr); }
+
+    void* untaggedExecutableAddress() const { return untagCodePtr<void*, tag>(m_ptr); }
+
+    explicit operator bool() const { return !!m_ptr; }
+    bool operator!() const { return !m_ptr; }
+
+    bool operator==(const TypedFunctionPtr& other) const { return m_ptr == other.m_ptr; }
+    bool operator!=(const TypedFunctionPtr& other) const { return m_ptr != other.m_ptr; }
+
+    TypedFunctionPtr& operator=(Ptr ptr)
+    {
+        m_ptr = encode(ptr);
+        return *this;
+    }
+
+    template<PtrTag otherTag>
+    TypedFunctionPtr& operator=(const TypedFunctionPtr<otherTag, Out, In...>& other)
+    {
+        m_ptr = encode(other.get());
+        return *this;
+    }
+
+    TypedFunctionPtr& operator=(std::nullptr_t)
+    {
+        m_ptr = nullptr;
+        return *this;
+    }
+
+protected:
+    enum AlreadyTaggedValueTag { AlreadyTaggedValue };
+    TypedFunctionPtr(AlreadyTaggedValueTag, void* ptr)
+        : m_ptr(bitwise_cast<Ptr>(ptr))
+    {
+        assertIsNullOrTaggedWith<tag>(ptr);
+    }
+
+    ALWAYS_INLINE static constexpr Ptr encode(Ptr ptr)
+    {
+        if constexpr (tag == CFunctionPtrTag)
+            return ptr;
+        if (!ptr)
+            return nullptr;
+        assertIsNullOrCFunctionPtr(ptr);
```

Comment:
we should assert this at the top of the function

Reply:
I cannot assert this at the top because it breaks usage of this function as `constexpr`.  I'll change it to `assertIsCFunctionPtr` there.